src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs

```diff
@@ -44,10 +44,14 @@ namespace System.IO.Enumeration
                 // We know it's a directory.
                 isDirectory = true;
             }
-            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK)
-                && Interop.Sys.Stat(entry.FullPath, out Interop.Sys.FileStatus targetStatus) >= 0)
+            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK
+                || directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
+                && (Interop.Sys.Stat(entry.FullPath, out Interop.Sys.FileStatus targetStatus) >= 0
+                    || Interop.Sys.LStat(entry.FullPath, out targetStatus) >= 0))
```

Comment:
I get that it's difficult to add tests for the UNKNOWN part of this. But could we add a test that fails with the missing LStat?

Reply:
> I get that it's difficult to add tests for the UNKNOWN part of this. But could we add a test that fails with the missing LStat?

I'm not sure how- do you have any ideas?

Reply:
If InodeType == DT_LNK and the Stat fails, is there ever a situation where isDirectory would then be set to true, where the LStat would then say that the Mode is S_IFDIR?  If no, then we probably can't come up with a test for it, and we could even consider restructuring to avoid calling LStat in that case.  If yes, then we could just write a test that gets into that condition.

Reply:
> If InodeType == DT_LNK and the Stat fails, is there ever a situation where isDirectory would then be set to true, where the LStat would then say that the Mode is S_IFDIR?

I'm not sure. :/ I'm trying to play it safe and match the semantics of https://github.com/dotnet/corefx/commit/59da18fc926397f6674e1de86e296ca6e0dfe067 as close as possible.

Reply:
Yeah, I agree with playing it safe for now and not changing the logic further now.  My question was more to guide whether there's anything to test there or whether that set of branches is actually dead code we couldn't possibly hope to write a unit test for.

Reply:
@JeremyKuhne @stephentoub I think it doesn't makes sense to call lstat here. Calling lstat after stat is a way of getting info about the link when the target doesn't exist. This code won't do anything (links are links, not directories). Does that make sense?

Reply:
@tmds, that's what I was getting at. But it does make sense for UNKNOWN. And we were previously calling lstat for both. The goal is to fix the regression at this point, not eliminate a nop.

Reply:
I don't understand. Why does calling lstat make sense for UNKNOWN? lstat wasn't called before this PR.

Reply:
It was.  In 2.0 we used DirectoryExists which called FileExists which did:
https://github.com/dotnet/corefx/blob/23ce8babbfe7a38b39f0719678de61c257acc61f/src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs#L445-L446
That regressed when the changes were made in 2.1 to improve enumeration perf.

Reply:
It's possible it doesn't make sense for UNKNOWN either, if the only way stat could possibly fail for UNKNOWN but lstat could succeed is if it's actually a symlink.  That can be explored being removed for 2.2.

Reply:
Ah ok. Still I think there is no need to call lstat.

When we want to check if an entry is a file and stat fails, we need to call lstat to check if it is a broken link. If that is the case, then by convention on Unix, we assume it is a file.

In this case, we are enumerating and we know something exists. At the point we call lstat we are already assuming it is a file (isDirectory is set to false).

Reply:
@tmds thanks for the input here. I'll yank the lstat call post 2.1 so it has time to bake, and loop you in on the PR (i.e., right after I port this to the 2.1 release branch).

If you have any other input on the Unix implementation in System.IO.FileSystem, I'd love to hear it.